protocol/messenger_communities.go

```diff
@@ -532,6 +535,87 @@ func (m *Messenger) updateCommunitiesActiveMembersPeriodically() {
 	}()
 }
 
+func (m *Messenger) HandleCommunityGrantUpdated(state *ReceivedMessageState, update *protobuf.CommunityGrantUpdated, statusMessage *v1protocol.StatusMessage) error {
+	community, err := m.communitiesManager.GetByID(update.CommunityId)
+	if err != nil {
+		return err
+	}
+
+	if community == nil {
+		return communities.ErrOrgNotFound
+	}
+
+	return m.communitiesManager.HandleCommunityGrant(community, update.Grant, update.Clock)
+}
+
+func (m *Messenger) publishGrantForCommunityMemeber(community *communities.Community, memberKey string) error {
+	memberPubKey, err := common.HexToPubkey(memberKey)
+	if err != nil {
+		return err
+	}
+
+	grant, err := community.BuildGrant(memberPubKey, "")
+	if err != nil {
+		return err
+	}
+
+	update := &protobuf.CommunityGrantUpdated{
+		Clock:       uint64(time.Now().Unix()),
+		CommunityId: community.ID(),
+		Grant:       grant,
+	}
+
+	payload, err := proto.Marshal(update)
+	if err != nil {
+		return err
+	}
+
+	rawMessage := &common.RawMessage{
+		Payload:             payload,
+		Sender:              community.PrivateKey(),
+		SkipEncryptionLayer: true,
+		MessageType:         protobuf.ApplicationMetadataMessage_COMMUNITY_GRANT_UPDATED,
+		PubsubTopic:         shard.DefaultNonProtectedPubsubTopic(),
+	}
+
+	_, err = m.sender.SendPrivate(context.Background(), memberPubKey, rawMessage)
+	return err
+}
+
+func (m *Messenger) updateGrantsForControlledCommunities() {
+	// Resend grants for controlled communities every hour
+	ticker := time.NewTicker(grantUpdateInterval)
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				controlledCommunities, err := m.communitiesManager.Controlled()
+				if err != nil {
+					m.logger.Error("failed fetch controlled communities for grants update", zap.Error(err))
+				}
+
+				for _, community := range controlledCommunities {
+					// Skip unencrypted communities and non-control node communities
+					if !community.IsControlNode() || community.Encrypted() {
+						continue
+					}
+
+					for memberKey := range community.Members() {
+						err = m.publishGrantForCommunityMemeber(community, memberKey)
```

Comment:
Could be quite an expensive thing. Every hour there will be a spike of waku usage, if you have many community members.
In future it would be nice to distribute the message sending across this hour. Smth like `delayPublishMessage(random(0-1hour), ...)`.

Reply:
Good idea! For now i would like to adjust timings

Reply:
Yeah, that would be expensive indeed. We could construct group message instead - a message containing all grants. We do the same for encryption keys, please see: `BuildHashRatchetReKeyGroupMessage`